README.md
tel.js - An input[tel] directive for AngularJS.
=================

A filter and a directive for input type 'tel' that formats and parses international phone numbers.
Based on metadata from libphonenumber.

Demo: http://michaelkrog.github.io/tel.js/

#FILTER#

###Usage###

####In HTML Template Binding####
```html
{{phonenumber_expression | telephone}}
```

####In Javascript####
```javascript
$filter('telephone')(phonenumber, mode, defaultAreaCode, returnObject);
```

###Arguments###

Param                      | Type      | Details
---------------------------|-----------|----------------------
phonenumber                | string    | Phonenumber to format.
mode (optional)            | string    | 'e164'(international) or 'national' format. Default 'e164'. 
defaultAreaCode (optional) | string    | If specified phonenumbers from this area will be formatted even without an areacode.
returnObject (optional)    | boolean   | If true the result will be an object of the form `{number:<number>,valid:<true|false>}`. If false or not specified the result will be the formatted result of the valid number, otherwise 'unspecified'.

#DIRECTIVE#

###Usage###
```html
<input type="tel" 
    ng-model="" 
    [international=""]
    [default-area-code=""]>
```

###Arguments###
Param                      | Type      | Details
---------------------------|-----------|----------------------
ngModel                    | string    | Assignable angular expression to data-bind to.
international (optional)   | string    | Wether number should be formatted as 'e164' or 'national' format. 'true' or 'false'. Default 'true'. 
defaultAreaCode (optional) | string    | If specified phonenumbers from this area will be formatted even without an areacode.


#Install#

You can install tel.js via bower like this:
```
bower install teljs --save
```

Alternatively you can download the full source as a ZIP file from [Github](https://github.com/michaelkrog/tel.js/archive/master.zip).


#FILE FORMAT#
The metadatalite.js file is taken directly from the libphonenumber repository. The format is defined by Google protobuf definitions and is therefore not very human readable. The following is a partial definition of the format - enough to build tel.js around the file.

__It is two Objects__

The whole file consists of 2 objects in the variables `i18n.phonenumbers.metadata.countryCodeToRegionCodeMap` and `i18n.phonenumbers.metadata.countryToMetadata`.

`i18n.phonenumbers.metadata.countryCodeToRegionCodeMap` holds a map of which countries use a given areacode.(fx. +45 is Denmark only, but +1 is a [NANPA](http://www.nanpa.com/) number and is used for USA, Canada and at least 20 other countries. The map refers to an array of countrycodes that uses a given area code having the main country as the first element in the array.

`i18n.phonenumbers.metadata.countryToMetadata` holds metadata for each country. Each metadata entry is retrieved by its countrycode(fx. 'DK'). The metadata entry is an array whici format is partly explained in the following table:

Field | Note
------|-----------
0     | unknown
1     | generalDesc: [unknown, unknown, nationalNumberPattern, possibleNumberPattern]
2     | fixedLine: [unknown, unknown, nationalNumberPattern, possibleNumberPattern]
3     | mobile: [unknown, unknown, nationalNumberPattern, possibleNumberPattern]
4     | tollfree: [unknown, unknown, nationalNumberPattern, possibleNumberPattern]
5     | premiumrat: [unknown, unknown, nationalNumberPattern, possibleNumberPattern]
6     | sharedCost: [unknown, unknown, nationalNumberPattern, possibleNumberPattern]
7     | personalNumber: [unknown, unknown, nationalNumberPattern, possibleNumberPattern]
8     | voip: [unknown, unknown, nationalNumberPattern, possibleNumberPattern]
9     | ISO countrycode (fx. 'DK') 
10    | areacode (fx. 45 for DK, 46 for SE etc.)
11    | internation prefix
12    | national prefix
13    | preferred extension prefix
14    | unknown
15    | national prefix for parsing
16    | national prefix transform rule
17    | preferred international prefix
18    | same mobile and fixed line pattern (boolean)
19    | national number formats. Array with entries of the following format: [unknown, pattarn, format, [leadingdigits...], nationalPrefixFormattingRule, unknown, unknown]
20    | international number formats. Array with entries of the following format: [unknown, pattarn, format, [leadingdigits...]]
21    | pager: [unknown, unknown, nationalNumberPattern, possibleNumberPattern]
22    | main country for code
23    | leading digits
24    | no international dialing: [unknown, unknown, nationalNumberPattern, possibleNumberPattern]
25    | uan: [unknown, unknown, nationalNumberPattern, possibleNumberPattern]
26    | leading zero possible
27    | emergency
28    | voicemail: [unknown, unknown, nationalNumberPattern, possibleNumberPattern]




CHANGELOG.md
## 1.0.11

* Ensured directive and filter are minsafe. 

## 1.0.10

* Fixed NANPA numbers outside US not being validated and formatted. (#14).

## 1.0.9

* Fixed model value having an extra plus added. (#11).

## 1.0.8

* No longer sets invalid on an empty value.(#9, #10).

## 1.0.7

* Fixed Javascript errors if model not initially defined.(#8).

## 1.0.6

* Isolate scope removed(#7).

## 1.0.5

* metadatalite.js updated to newest revision from libphonenumber(#6).


## 1.0.4

* Filter now returns raw input on invalid numbers instead of empty string(#5).


src/tel.js
/*global angular, goog, i18n */
/*jslint browser: true */
/* Fallback if google protocols is not defined - START*/
if (window.goog === undefined) {
    goog = {
        provide: angular.noop
    };

    i18n = {
        phonenumbers: {
            metadata: {
            }
        }
    };
}
/* Fallback if google protocols is not defined - END*/

var teljs = angular.module('teljs',[]);
teljs.trimNumber = function(value) {
    if (value) {
        return value.replace(/[\+\s\-\(\)]/g, '');
    } else {
        return value;
    }
};

teljs.filter('telephone', [function() {

    var countryCodes = i18n.phonenumbers.metadata.countryCodeToRegionCodeMap;
    var countryMetaData = i18n.phonenumbers.metadata.countryToMetadata;
    var validRangeMap = {
        2: 'fixedLine',
        3: 'mobile',
        4: 'tollFree',
        5: 'premiumRate',
        6: 'sharedCost',
        7: 'personalNumber',
        8: 'voip',
        21: 'pager',
        24: 'noInternationalDialing',
        25: 'uan',
        28: 'voiceMail'
    };

    var regionsFromNumber = function(e164) {
        var regions, key, value;
        for(key in countryCodes) {
            value = countryCodes[key];
            var reg = new RegExp('^'+key),
            ok;

            ok = reg.exec(e164);

            if(ok) {
                regions = value;
                break;
            }
        }

        return regions;
    };

    var formatNumberForRegion = function(region, nationalNumber, mode, mainRegion) {
        var metaData = countryMetaData[region],
            metaDataMain = countryMetaData[mainRegion],
            countryCode = metaData[10],
            nationalPrefix = metaData[12],
            nationalFormats = metaDataMain[19],
            internationalFormats = metaDataMain[20],
            number, international = (mode === 'e164'),
            numberFormats = international && internationalFormats ? internationalFormats : nationalFormats,
            entry, i, matchNumber, matchLeadingDigits, validRange = false, range;

        if(numberFormats) {

            if(nationalPrefix && nationalNumber.substr(0, nationalPrefix.length) === nationalPrefix) {
                nationalNumber = nationalNumber.substr(nationalPrefix.length);
            }

            for (var key in validRangeMap) {
                entry = metaData[key];
                if(new RegExp('^('+entry[2]+')$').exec(nationalNumber)) {
                    validRange = true;
                    range = validRangeMap[key];
                    break;
                }
            }

            if(validRange) {
                for(i = 0;i<numberFormats.length;i++) {
                    entry = numberFormats[i];
                    matchNumber = new RegExp('^('+entry[1]+')$').exec(nationalNumber);

                    if(!angular.isDefined(entry[3])) {
                        matchLeadingDigits = true;
                    } else if(angular.isString(entry[3])) {
                        matchLeadingDigits = new RegExp('^('+entry[3]+')').exec(nationalNumber);
                    } else if(angular.isArray(entry[3])) {
                        angular.forEach(entry[3], function(lead) {
                            var result = new RegExp('^('+lead+')').exec(nationalNumber);
                            if(result) {
                                matchLeadingDigits = true;
                            }
                        });
                    }

                    if(matchNumber && matchLeadingDigits) {
                        var format = entry[2];
                        if(international) {
                            number = nationalNumber;
                            number = number.replace(new RegExp(entry[1]), format);
                            number = '+' + countryCode + ' ' + number;
                        } else {
                            number = nationalNumber;
                            if(nationalPrefix) {
                                number = nationalPrefix + '' + number;
                            }
                            number = number.replace(new RegExp(entry[1]), format);
                        }
                        break;
                    }
                }
            }
        }
        return number;
    };

    var formatNumber = function(number, mode) {
        var regions = regionsFromNumber(number), region, i, countryCode, nationalPrefix, nationalNumber, result, mainRegion;
        if(regions && regions.length > 0) {
            mainRegion = regions[0];

            for(i=0;i<regions.length;i++) {
                region = regions[i];
                countryCode = countryMetaData[region][10];
                nationalPrefix = countryMetaData[region][12];
                nationalNumber = number.substr(countryCode.toString().length);
                if(nationalPrefix && nationalNumber.substr(0, nationalPrefix.length) !== nationalPrefix) {
                    nationalNumber = nationalPrefix + '' + nationalNumber;
                }
                result = formatNumberForRegion(region, nationalNumber, mode, mainRegion);
                if (result) {
                  break;
                }
            }
        }
        return result;
    };

    var wrapResult = function(input, number, returnObject) {
        var result;

        if(returnObject === true) {
            if(!number) {
                result = {
                    number: input,
                    valid: false
                };
            } else {
                result = {
                    number: number,
                    valid: true
                };
            }
        } else {
            if(!number) {
                result = input;
            } else {
                result = number;
            }
        }
        return result;
    };

    return function(input, mode, defaultAreaCode, returnObject) {

        var trimmedNumber, defaultGeneratedNumber, number;
        mode = mode ? mode : 'e164';
        trimmedNumber = teljs.trimNumber(input);
        if(defaultAreaCode && defaultAreaCode !== '') {
            defaultGeneratedNumber = defaultAreaCode + '' + trimmedNumber;
            number = formatNumber(defaultGeneratedNumber, mode);
            if (number) {
                return wrapResult(input, number, returnObject);
            }
        }

        number = formatNumber(trimmedNumber, mode);
        return wrapResult(input, number, returnObject);
    };
}]);

teljs.directive('input', ['$filter', function($filter) {
    return {
        restrict: 'E', // only activate on element attribute
        require: '?ngModel', // get a hold of NgModelController
        link: function(scope, element, attrs, ngModel) {
            if (attrs.type !== 'tel') {
                return;
            }
            scope.international = attrs.international;
            scope.defaultAreaCode = attrs.defaultAreaCode;

            if(scope.international !== 'false') {
                scope.mode = 'e164';
            } else {
                scope.mode = 'national';
            }

            element.on('blur', function() {
                if(ngModel.$valid) {
                    ngModel.$setViewValue(scope.formatNumber(ngModel.$modelValue));
                    ngModel.$render();
                }
            });

            scope.doFormatNumber = function(number, mode) {
                return $filter('telephone')(number, mode, scope.defaultAreaCode, true);
            };

            scope.formatNumber = function(value) {
                var result;

                if(!angular.isDefined(value) || value === '') {
                    return '';
                }

                result = scope.doFormatNumber(value, scope.mode);
                if(!result.valid) {
                    result.number = value;
                    ngModel.$setValidity('phoneNumber', false);
                }
                return result.number;
            };

            scope.parseNumber = function(value) {
                var formatResult;
                value = value ? teljs.trimNumber(value) : value;
                formatResult = scope.doFormatNumber(value, 'e164');

                if (!formatResult.valid && (value === '' || value === undefined)) {
                    formatResult.valid = true;
                    formatResult.number = '';
                }

                ngModel.$setValidity('phoneNumber', formatResult.valid);
                return formatResult.number !== '' ? '+' + teljs.trimNumber(formatResult.number) : '';
            };

            ngModel.$formatters.push(scope.formatNumber);
            ngModel.$parsers.push(scope.parseNumber);

        }
    };
}]);
